### PR TITLE
Use glaze to parse hyprctl json output, other improvements

### PR DESCRIPTION
**Pull request by zjeffer**
#### Describe your PR, what does it fix/add?
* Removed the very ugly manual parsing of the output of `hyprctl plugin list | grep Plugin`, by replacing it with a simple glaze  parse of the json output of `hyprctl plugin list -j` in PluginManager.cpp
* While I was busy, I noticed the patterns `folder + "/" + filename` and `folder + "/filename.ext"` came up a lot. I really dislike this so I replaced it with the std::filesystem `/` operator  (in a separate commit)
* I also added a helper method to get the state.toml files for every plugin, because the exact same logic was used 4 times in the same file. I extracted the logic into a new helper function "getPluginstates" in DataState.
* Also took care of some clang-tidy suggestions, for example `find` was being called with a single-character string `"x"` instead of just a character `'x'`.

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
I'm planning on replacing more of these hyprctl calls with json parsing, like we discussed in this PR: https://github.com/hyprwm/Hyprland/pull/7152#issue-2445664270

#### Is it ready for merging, or does it need work?
Ready to merge

### CODE REVIEW

**Comment by vaxerski** [comment]
can you use something that doesn't suck as horrendously as nlohmann's json

**Comment by zjeffer** [comment]
What would you prefer? [Here](https://github.com/hyprwm/Hyprland/pull/7152#issuecomment-2266121214) you said nlohmann::json is ok, and I'd get banned for using boost::json, and you said you don't know jsoncpp. 

rapidjson?

**Comment by vaxerski** [comment]
yeh I was a bit quick on that one back then. Nlohmann's json is ungodly slow. Check out https://github.com/stephenberry/glaze

**Comment by zjeffer** [comment]
done, let me know if you want me to use git submodules instead of cmake's fetchcontent

File: hyprpm/CMakeLists.txt [hunk 1]
```diff
@@ -13,6 +13,17 @@ pkg_check_modules(hyprpm_deps REQUIRED IMPORTED_TARGET tomlplusplus hyprutils>=0
...
+# glaze
+include(FetchContent)
+FetchContent_Declare(
+  glaze
+  GIT_REPOSITORY https://github.com/stephenberry/glaze.git
+  GIT_TAG v4.2.2
+  GIT_SHALLOW TRUE
+)
+FetchContent_MakeAvailable(glaze)
+target_link_libraries(hyprpm PRIVATE glaze::glaze)
```
Comment (vaxerski) [inline hyprpm/CMakeLists.txt:25]: glaze is in the AUR: https://repology.org/project/glaze/versions

Do not do this, just add a dep in pkg-config.

Hyprland is a big enough project to not worry about glaze not being widely packaged, because we can force packagers to package it. :)

Comment (zjeffer) [reply hyprpm/CMakeLists.txt]: How would you fix the CI/CD, then? Should I add a step to the setup_base action to compile glaze from source?

In the latest commit, I tried adding it as a submodule which works for cmake but not yet for meson, haven't figured out how to do that yet. But I guess I'll just remove it and do it with pkg-config instead.

Comment (zjeffer) [reply hyprpm/CMakeLists.txt]: Had to use find_package cause glaze doesn't supply a .pc file.

Comment (vaxerski) [reply hyprpm/CMakeLists.txt]: > How would you fix the CI/CD, then? Should I add a step to the setup_base action to compile glaze from source?

CI/CD already pulls from git some dependencies. Just add a new one.

> In the latest commit, I tried adding it as a submodule which works for cmake but not yet for meson, haven't figured out how to do that yet. But I guess I'll just remove it and do it with pkg-config instead.

you have to ping @fufexan 3 times at 2am and he will appear behind you and fix meson

Comment (zjeffer) [reply hyprpm/CMakeLists.txt]: meson seems to work now with the system-installed glaze.
But I'm having issues with the Nix build and I have 0 experience with that. Who do I conjure up for that? :smile:

Comment (vaxerski) [reply hyprpm/CMakeLists.txt]: fufexan as well

**Comment by fufexan** [comment]
I'm not sure whether the arch package for glaze contains a .pc file, but the Nixpkgs one only includes cmake files, so I had to modify the lookup method in meson.build.

**Comment by zjeffer** [comment]
Thanks! Ready to review now.

File: hyprpm/src/core/DataState.cpp [hunk 2]
```diff
@@ -72,17 +90,10 @@ bool DataState::pluginRepoExists(const std::string& urlOrName) {
 
     const auto PATH = getDataStatePath();
 
-    for (const auto& entry : std::filesystem::directory_iterator(PATH)) {
-        if (!entry.is_directory() || entry.path().stem() == "headersRoot")
-            continue;
-
-        if (!std::filesystem::exists(entry.path().string() + "/state.toml"))
```
Comment (vaxerski) [inline hyprpm/src/core/DataState.cpp:79]: why did you remove all these safeguards?

Comment (vaxerski) [reply hyprpm/src/core/DataState.cpp:79]: touche

**Review by vaxerski — APPROVED** [review]
Thanks!

**Comment by Tchekda** [comment]
Hello @zjeffer since your PR, I am having issues building my nixos config with Hyprland. 
Here is the error I got when using today's rev as my flake input:
```
error:
       … while calling the 'head' builtin
         at /nix/store/8vz84mqgnm1gz5yk7hgnnb5gir5hjxas-source/lib/attrsets.nix:1574:11:
         1573|         || pred here (elemAt values 1) (head values) then
         1574|           head values
             |           ^
         1575|         else

       … while evaluating the attribute 'value'
         at /nix/store/8vz84mqgnm1gz5yk7hgnnb5gir5hjxas-source/lib/modules.nix:846:9:
          845|     in warnDeprecation opt //
          846|       { value = addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          847|         inherit (res.defsFinal') highestPrio;

       … while evaluating the option `system.build.toplevel':

       … while evaluating definitions from `/nix/store/8vz84mqgnm1gz5yk7hgnnb5gir5hjxas-source/nixos/modules/system/activation/top-level.nix':

       … while evaluating the option `warnings':

       … while evaluating definitions from `/nix/store/8vz84mqgnm1gz5yk7hgnnb5gir5hjxas-source/nixos/modules/system/boot/systemd.nix':

       … while evaluating the option `systemd.services.home-manager-tchekda.serviceConfig':

       … while evaluating definitions from `/nix/store/8vz84mqgnm1gz5yk7hgnnb5gir5hjxas-source/flake.nix':

       … while evaluating the option `home-manager.users.tchekda.home.activation.checkFilesChanged.data':
       … while evaluating definitions from `/nix/store/0s5bx5a9565ialdjc688dar6lpz9d6bz-source/modules/files.nix':

       … while evaluating the option `home-manager.users.tchekda.home.file."/home/alastyn/.config/hypr/hyprland.conf".onChange':

       … while evaluating definitions from `/nix/store/0s5bx5a9565ialdjc688dar6lpz9d6bz-source/modules/misc/xdg.nix':

       … while evaluating the option `home-manager.users.tchekda.xdg.configFile."hypr/hyprland.conf".onChange':

       … while evaluating definitions from `/nix/store/0s5bx5a9565ialdjc688dar6lpz9d6bz-source/modules/services/window-managers/hyprland.nix':

       … while evaluating the option `home-manager.users.tchekda.wayland.windowManager.hyprland.package':

       … while evaluating definitions from `/nix/store/xhlwzy6rs7yh8cabim8iciz26ammkr5c-source/modules/home/wayland/hyprland':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: evaluation aborted with the following error message: 'lib.customisation.callPackageWith: Function called without required argument "glaze" at /nix/store/bnmimbjj61hp0aigzrkwpms8b9cyk7yn-source/nix/default.nix:15, did you mean "blaze", "glade" or "clazy"?'
```

When forcing to the rev before your PR `github:hyprwm/Hyprland/9f933da1c502989cadf7696971aa376d65847b95` it works well. 
Can you please look into it as I couldn't find the reason.
Thank you

cc @AlastynW

**Comment by zjeffer** [comment]
I know nothing about nix so I can't really help here, but don't you simply need to add glaze to your nix config? It's a new dependency.

**Comment by fufexan** [comment]
Glaze was already added to Hyprland itself. @Tchekda do you override Hyprland's Nixpkgs input?
